core/controllers/learner_dashboard.py

```diff
@@ -0,0 +1,206 @@
+            learner_progress_services.remove_none_entities_from_list(
+                self.user_id,
+                exp_services.get_exploration_summaries_matching_ids(
+                    completed_exploration_ids),
+                completed_exploration_ids, 'complete', 'explorations'))
+
+        outdated_incomplete_collection_ids = (
+            learner_progress_services.get_all_incomplete_collection_ids(
+                self.user_id))[:]
+        completed_collection_ids = (
+            learner_progress_services.get_all_completed_collection_ids(
+                self.user_id))
+        updated_incomplete_collection_ids = (
+            learner_progress_services.get_all_incomplete_collection_ids(
+                self.user_id))
+
+        completed_to_incomplete_collections = []
+        # Get the collections that were updated to have new explorations.
+        completed_to_incomplete_collection_ids = (
+            set(outdated_incomplete_collection_ids).symmetric_difference(
+                updated_incomplete_collection_ids))
+
+        collection_titles_and_categories = (
+            collection_services.get_collection_titles_and_categories(
+                list(completed_to_incomplete_collection_ids)))
+        for collection_id in collection_titles_and_categories:
+            completed_to_incomplete_collections.append(
+                collection_titles_and_categories[collection_id]['title'])
+
+        completed_collection_summaries, number_of_deleted_completed_collections = ( # pylint: disable=line-too-long
+            learner_progress_services.remove_none_entities_from_list(
+                self.user_id,
+                collection_services.get_collection_summaries_matching_ids(
+                    completed_collection_ids),
+                completed_collection_ids, 'complete', 'collections'))
+
+        updated_incomplete_collection_ids = (
+            learner_progress_services.get_all_incomplete_collection_ids(
+                self.user_id))
+        incomplete_collection_summaries, number_of_deleted_incomplete_collections = ( # pylint: disable=line-too-long
+            learner_progress_services.remove_none_entities_from_list(
+                self.user_id,
+                collection_services.get_collection_summaries_matching_ids(
+                    updated_incomplete_collection_ids),
+                updated_incomplete_collection_ids, 'incomplete', 'collections'))
+
+        completed_exp_summary_dicts = (
+            summary_services.get_displayable_exp_summary_dicts(
+                completed_exploration_summaries))
+
+        incomplete_exp_summary_dicts = (
+            summary_services.get_displayable_exp_summary_dicts(
+                incomplete_exploration_summaries))
+
+        completed_collection_summary_dicts = []
+        incomplete_collection_summary_dicts = []
+
+        for collection_summary in completed_collection_summaries:
```

Comment:
There's still lots of logic and manual assembly here. I suggest finding a way to push it back into the services layer so that you can just do something like get_completed_collection_summary_dicts(). Alternatively you can also do, in the final dict:

`'completed_collections_list': [summary.to_dict() for summary in completed_collection_summaries]`

which is probably preferable -- it makes sense for the to_dict() method to be on the summary object. Ditto below.